styx-scheduler-service/src/main/java/com/spotify/styx/api/SchedulerResource.java

```diff
@@ -82,11 +91,42 @@ public SchedulerResource(
         Route.with(
             em.response(WorkflowInstance.class),
             "POST", BASE + "/trigger",
-            rc -> this::triggerWorkflowInstance)
+            rc -> this::triggerWorkflowInstance),
+        Route.with(
+            em.response(Workflow.class),
+            "POST", BASE + "/workflows",
+            rc -> this::createOrUpdateWorkflow),
+        Route.with(
+            em.serializerResponse(ByteString.class),
+            "DELETE", BASE + "/workflows/<cid>/<wfid>",
+            rc -> deleteWorkflow(rc.pathArgs().get("cid"), rc.pathArgs().get("wfid")))
     )
+
         .map(r -> r.withMiddleware(Middleware::syncToAsync));
   }
 
+  private Response<ByteString> deleteWorkflow(String cid, String wfid) {
+    WorkflowId workflowId = WorkflowId.create(cid, wfid);
+    Optional<Workflow> workflowOpt = null;
+    try {
+      workflowOpt = storage.workflow(workflowId);
+    } catch (IOException e) {
+      return Response
+          .forStatus(Status.INTERNAL_SERVER_ERROR.withReasonPhrase("Error in internal storage"));
+    }
+    if (!workflowOpt.isPresent()) {
+      return Response.forStatus(Status.NOT_FOUND.withReasonPhrase("Workflow does not exist"));
+    }
+    workflowRemoveListener.accept(workflowOpt.get());
+    return Response.forStatus(Status.NO_CONTENT);
+  }
+
+  private Response<Workflow> createOrUpdateWorkflow(Workflow workflow) {
+    workflowChangeListener.accept(workflow);
```

Comment:
It's better to have a big `TODO` here: validate the workflow and throw something more reasonable than 500 if `accept` failed with `RuntimeException`. This can be address in a different PR to keep this one clean.

Reply:
We should validate the workflow as much as possible, and return 4XX for bad requests, but if something fails after validation I think that's an internal error that should be surfaced as a 500.

Reply:
There was a corner case that invalid workflows are surfaced later when parsing cron syntax to get next natural trigger. We can address that in a separated PR.